pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -3487,6 +3487,156 @@ func TestReconcileWithWhenExpressionsWithTaskResults(t *testing.T) {
 	}
 }
 
+func TestReconcileWithWhenExpressionsScopedToTask(t *testing.T) {
+	names.TestingSeed()
+	ps := []*v1beta1.Pipeline{tb.Pipeline("test-pipeline", tb.PipelineNamespace("foo"), tb.PipelineSpec(
+		tb.PipelineTask("a-task", "a-task"),
+		tb.PipelineTask("b-task", "b-task",
+			tb.PipelineTaskWhenExpression("$(tasks.a-task.results.aResult)", selection.In, []string{"aResultValue"}),
+			tb.PipelineTaskWhenExpression("aResultValue", selection.In, []string{"$(tasks.a-task.results.aResult)"}),
+		),
+		tb.PipelineTask("c-task", "c-task",
+			tb.PipelineTaskWhenExpression("$(tasks.a-task.results.aResult)", selection.In, []string{"missing"}),
+		),
+		tb.PipelineTask("d-task", "d-task", tb.RunAfter("c-task")),
```

Comment:
i'm a bit confused by the setup of this test (maybe some comments might help?)

it seems like we're making a graph of:

a-task runs on its own
b-task depends on a-task, is guarded with a when expression based on a-task which will always be true
c-task depends on a-task, is guarded with a when expression based on a-task which will always be falst
d-task runs after c-task

we're setting this up so that a-task has already run, therefore b-task and c-task will get evaluated next - b-task should run, and c-task should be skipped

Since this test turns on `scope-when-expressions-to-task`, i would expect that skipping c-task would have no impact d-task, i.e. it would start running at the same time as b-task

(with `scope-when-expressions-to-task` off, the default, id expect d-task to get skipped also) 

so after that wall of text haha, im wondering:
1. am i understanding the setup correctly? (and maybe can we add some comments to the test to make it more clear)
2. why isn't d-task also running after the reconcile?

Reply:
> i'm a bit confused by the setup of this test (maybe some comments might help?)
> 
> it seems like we're making a graph of:
> 
> a-task runs on its own
> b-task depends on a-task, is guarded with a when expression based on a-task which will always be true
> c-task depends on a-task, is guarded with a when expression based on a-task which will always be false
> d-task runs after c-task
> 

yes, that's the correct set up of the test

> we're setting this up so that a-task has already run, therefore b-task and c-task will get evaluated next - b-task should run, and c-task should be skipped
> 

that's right, then after b-task and c-task are done, d-task will be evaluated in the next reconcile loop

> Since this test turns on `scope-when-expressions-to-task`, i would expect that skipping c-task would have no impact d-task, i.e. it would start running at the same time as b-task
> 

d-task can't run at the same time as b-task -- it has to run in the next loop (after b-task and c-task are done) because of how we traverse the pipeline graph

regardless of the when expressions being scoped to task, the execution order of tasks is still enforced and we won't look at d-task until the next time getNextTasks is called in the next reconcile loop 
https://github.com/tektoncd/pipeline/blob/c8dc797cf5a6f11f90cb742d014470a444fcdc60/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L219-L240

> (with `scope-when-expressions-to-task` off, the default, id expect d-task to get skipped also)
> 

yes, that's right

> so after that wall of text haha, im wondering:
> 
> 1. am i understanding the setup correctly? (and maybe can we add some comments to the test to make it more clear)

good idea, added some comments to the test and also simplified it to need only one reconcile loop to make it clearer

> 2. why isn't d-task also running after the reconcile?

in this reconcile, only b-task and c-task will run -- d-task is supposed to run in the next

Reply:
updated the test such that we have:
- a-task runs on its own, is guarded with when expressions that evaluate to false so is skipped
- b-task depends on a-task based on runafter, is executed because it runs after a-task and when expressions are scoped to task
- c-task depends on a-task based on result, is attempted because when expressions are scoped to task but then get skipped because of missing result references from a-task

Reply:
> d-task can't run at the same time as b-task -- it has to run in the next loop (after b-task and c-task are done) because of how we traverse the pipeline graph

that doesnt sound right to me - is this the way we evaluate other concurrent tasks? e.g. if a pipeline has 2 tasks, `a` and `b`, and they have no dependencies or `runAfter`, they both run simultaneously - do they not get both created in the next reconcile loop, or do we do them one at a time? 

this line makes me think we start all the tasks that are available to run:

https://github.com/tektoncd/pipeline/blob/8907e408c73480482c618d08eeaaefe0bd4aac68/pkg/reconciler/pipelinerun/pipelinerun.go#L637

i dont think that when expressions should make that behavior any different, and i would count that as a bug in when expressions if they do (how can we even guarantee in that case that the pipelinrun would immediately be reconciled again?)

Reply:
@jerop and i spent some time talking this through and i take back what i said about the behavior being a bug XD sorry for jumping to that conclusion @jerop ! 🙏 

Sounds like there are 2 different and valid approaches we could use with different trade offs. Imagine a Pipeline with 100 tasks: they run sequentially (using runAfter) but the first 99 have when expressions that evaluate to false and they should be skipped:

```
    1 (when false)
    |
    V
    2 (when false)
    |
    V
    3 (when false)
    |
    V
    4...99 (when false)
    |
    V
    100
```

The question is: what happens in the first reconcile loop?

1. I assumed that the entire graph would be evaluated and at the end of the first reconcile loop, 100 would be run
2. The implementation here would look at the next task to run, which on the first reconcile would be 1, evaluate it, determine it should be skipped, and complete. The next reconcile would look at 2, and so on.

(1) assumes the job of the reconcile loop is to determine which task should _run_ next
(2) assumes the job of the reconcile loop is to determine which task should be _evaluated_ next

I think (2) is simpler to implement (esp given our existing code) b/c going with (1) would mean we have to do some kind of "evaluate until done" looping within the reconcile loop that we've never had to do before.

But I think (2) is less efficient b/c running the above pipeline would take 100 separate reconciles (including calls to the API server, to etcd, and the being notified by the API server that we need to evaluate again). (BUT (1) might potentially block reconciling of other pipelines by doing a loop over 99 when expressions in one reconcile loop?)

@jerop  and I think it makes sense to go with (2) for now and address possible efficiency issues later if needed (🤞  hopefully no one is running a Pipeline like the above XD)

requests from me:
- Verify that the reconcile loop will be triggered again after a Task is evaluated as skipped (i.e. that the change we are making in the reconcile loop will trigger another reconcile) - if not, we'll need to use the approach where we traverse all the skipped tasks in one loop
- Add something to our [developer docs](https://github.com/tektoncd/pipeline/tree/main/docs/developers) explaining the tradeoff made here (e.g. the simplicity in using multiple reconciles vs getting everything done in one reconcile)
- I liked the test case with d-task and think it's worth adding back XD it stimulated a great conversation

Reply:
@bobcatfish you were actually right - turns out that what we actually do is (1) where the reconcile loop determines which task would run next

followed up with @pritidesai about the `successfulOrSkippedDAGTasks` function which enables (1) to happen: https://github.com/tektoncd/pipeline/blob/a728ce3c3d717fe52ff0b1eb71d357460498115c/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L480-L490

so what happened in the initial test case? 
- `a-task` runs on its own
- `b-task` depends on `a-task` (based on `runAfter`), is guarded with a `when` expression which will always be true
    - is executed successfully
- `c-task` depends on `a-task` (based on a `Result`), is guarded with a `when` expression with a missing `Result`
    - is skipped because of a missing `Result`
    - test case was misconstructed because `a-task` ran successfully but somehow didn't produce that `Result`
- `d-task` runs after `c-task`
    - is skipped because its parent task `c-task` was skipped because of a missing `Result`

I'd thought what we did was (2) where the reconcile loop determines which task would be evaluated next, and that that was the reason `d-task` was not executed but it was actually because of the above reason

I've added expanded the test case to cover the multiple scenarios with both ordering and resource dependent tasks